packages/muelu/test/scaling/MatvecKernelDriver.cpp

```diff
@@ -100,46 +100,34 @@ void print_crs_graph(std::string name, const V1 rowptr, const V2 colind) {
 }
 
 
+
 // =========================================================================
-// CuSparse Testing
+// MAGMASparse Testing
 // =========================================================================
-#if defined(HAVE_MUELU_CUSPARSE) && defined(HAVE_MUELU_TPETRA)
-#include <cuda_runtime.h>
-#include <cusparse.h>
-#include <cublas_v2.h>
-
-// The cuda examples provide a nice helper function that does all
-// of this error checking
-// helper_cuda.h
-// I am tempted to place it into our repo... but I'll need to read the license first.
-// Worst case, we could replicate here (the following is essentially part of the funcitonality
-// but they do it better.
-
-std::string cusparse_error(cusparseStatus_t code) {
-  switch(code) {
-  case CUSPARSE_STATUS_SUCCESS:
-    return std::string("CUSPARSE_STATUS_SUCCESS: Success");
-  case CUSPARSE_STATUS_NOT_INITIALIZED:
-    return std::string("CUSPARSE_STATUS_NOT_INITIALIZED: Library not initialized.");
-  case CUSPARSE_STATUS_ALLOC_FAILED:
-    return std::string("CUSPARSE_STATUS_ALLOC_FAILED: resources could not be allocated.");
-  case CUSPARSE_STATUS_INVALID_VALUE:
-    return std::string("CUSPARSE_STATUS_INVALID_VALUE: invalid parameters were passed (m,n,nnz<0).");
-  case CUSPARSE_STATUS_ARCH_MISMATCH:
-    return std::string("CUSPARSE_STATUS_ARCH_MISMATCH: the device does not support double precision (compute capability (c.c.) >= 1.3 required), symmetric/Hermitian matrix (c.c. >= 1.2 required), or transpose operation (c.c. >= 1.1 required).");
-  case CUSPARSE_STATUS_INTERNAL_ERROR:
-    return std::string("CUSPARSE_STATUS_INTERNAL_ERROR: Internal error");
-  case CUSPARSE_STATUS_MATRIX_TYPE_NOT_SUPPORTED:
-    return std::string("CUSPARSE_STATUS_MATRIX_TYPE_NOT_SUPPORTED: Matrix type not supported");
-  default:
-    return (std::string("ERROR: cuda_error got a return code it doesn't not understand") + std::to_string(code));
-  }
-}
+#if defined(HAVE_MUELU_MAGMASPARSE) && defined(HAVE_MUELU_TPETRA)
+#include <magma_v2.h>
+#include <magmasparse.h>
+template<typename Scalar, typename LocalOrdinal, typename GlobalOrdinal, typename Node>
+class MagmaSparse_SpmV_Pack {
+  typedef Tpetra::CrsMatrix<Scalar,LocalOrdinal,GlobalOrdinal,Node> crs_matrix_type;
+  typedef Tpetra::MultiVector<Scalar,LocalOrdinal,GlobalOrdinal,Node> vector_type;
+
+public:
+  MagmaSparse_SpmV_Pack (const crs_matrix_type& A,
+                      const vector_type& X,
+                            vector_type& Y)
+  {}
```

Comment:
It would be nicer for this class not to compile at all if Scalar is not one of the supported types.

Reply:
I'm not sure. This gets complicated, because we instantiate the driver itself for multiple ETI types.  The general idea is that the code will compile, but that it simply won't provide the runtime option.

What we really need, is a better way to express the different instantiations (it would speed up build times).

Reply:
Christian's way of doing it in kokkos-kernels makes sense.  You have a struct with a static method that says "try using `${TPL}` to solve the problem."  If `${TPL}` can't do it -- e.g., because it doesn't support the Scalar type -- the method returns false, and the caller has to try something else.